util/log.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import print_function
+import __builtin__
+import sys
+
+# used for looking up level from step=0 to error=4
+LEVEL_NAMES = 'step,debug,info,warn,error'.split(',')
+
+def set_log_levels(log_level_str):
+    '''
+    Sets the global log levels by parsing a log level configuration string.
+    Typically called at startup by main routine of a program (passing-in a command line value).
+    'log_level_str' - comma separated assignments of log levels to modules (e.g. "main=debug,persistence=step")
+        Modules are specified by their mudule IDs. 
+        Levels can be "step", "debug", "info", "warn" and "error".
+        If a part specifies only a level, this level will be used as the default one.
+    '''
+    global DEFAULT_LOG_LEVEL
+    global MODULE_LOG_LEVELS
+    DEFAULT_LOG_LEVEL = 2 # default is info
+    MODULE_LOG_LEVELS = {} # and no module specific levels
+    for part in log_level_str.split(','):
+        if part in LEVEL_NAMES:
+            DEFAULT_LOG_LEVEL = LEVEL_NAMES.index(part)
+        else:
+            module_level = part.split('=')
+            if len(module_level) == 2 and module_level[1] in LEVEL_NAMES:
+                MODULE_LOG_LEVELS[module_level[0]] = LEVEL_NAMES.index(module_level[1])
+    # persist settings in cross module global scope - other log-module imports will see those values
+    __builtin__.DEFAULT_LOG_LEVEL = DEFAULT_LOG_LEVEL
+    __builtin__.MODULE_LOG_LEVELS = MODULE_LOG_LEVELS
+
+class Logger(object):
+    '''
+    Class for logging messages to the console. Supports log levels and module based logging.
+    '''
+    def __init__(self, module_id, module_prefix):
```

Comment:
Can we rename `module_id` and `module_prefix` to `id` and `prefix` and use them as keyword arguments when instanciating the class? It makes it easier to read and follow IMHO: `Logger(id='...', prefix='...')` rather than `Logger('...', '...')`

Reply:
They are both mandatory and cannot default to something - sorry.

Reply:
You don't have to set a default value to use keyword arguments:

    >>> def foo(bar, baz):
    ...     print(bar, baz)
    ...
    >>> foo(baz="123", bar="abc")
    abc 123

Reply:
Ah - now I got it... done.